Analyze pathways

nuclearkatie opened the pull request:
This PR includes functions to analyze acquisition pathways. Testing is now at full coverage.

PR includes
- `trailmap/pathway_analysis.py`, a set of functions that conduct various analyses of acquisition pathways and/or the fuel cycle graph itself
- `tests/test_pathway_analysis.py`, testing for pathway analysis
- `tests/pa_data.py` data used in some of the tests
- and update to the README and CircleCI config file to note package requirements

nuclearkatie commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+    else:
+        print("Provided pathway has multiple edges. Flow calculation across \
+               multi-edged pathways is currently unsupported. Returning None")
+        return None
```
Note that I'm not supporting pathways that could be duplicated due to the existence of multiple edges between a pair of nodes

for example, the following graph
1 - 2 = 3 - 4
would result in two pathways,
- (1,2,3,4)
- (1,2,3,4)

They appear identical when listing the nodes of the pathway, but the (2,3) edge is different and may in fact represent different commodities. Right now, `acquisition_pathways` destroys duplicate paths because they're stored as a set. I raised this in issue #35, and I recommend that it be addressed in a future PR instead of this one because it will require changes to `acquisition_pathways` enumeration

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+
+        # check for capacity in edges (if no capacity, NetworkX will raise an
+        # exception) and if every edge has infinite flow, return inf
+        infinite_flow = True
```
Might be able to streamline/simplify all this by using `path_flow` as your logic variable and whether it `isinf()` as your test.

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+    return sc
+
+
+def check_if_sublist(path, set_of_steps):
```
What is the form of `set_of_steps`?

nuclearkatie replied:
should be a list, so calling it a set is a silly idea...

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+    pos = -1
+    sub_list = False
+
+    for i in range(len(path) - len(set_of_steps)+1):
+        if path[i] == set_of_steps[0]:
+            n = 1
+            while (n < len(set_of_steps) and path[i+n] == set_of_steps[n]):
+                n+=1
+            if n == len(set_of_steps):
+                sub_list = True
+                pos = i + len(set_of_steps)
+                break
```
Does this also do what you want? (might need to double check the "+1" logic)
```suggestion
num_steps = len(set_of_steps)

start = 0
end = start + num_steps + 1
while (end < len(path)) and (path[start:end] != set_of_steps):
    start += 1
    end = start + num_steps + 1
```

nuclearkatie replied:
I guess with this, you still need to implement an additional check to confirm if the while loop broke because it hit the nd of the path or because it found the full sublist. How about this instead?

```suggestion
        num_steps = len(list_of_steps)
    start = 0
    end = start + num_steps
    while end <= len(path):
        if path[start:end] == tuple(list_of_steps):
            return True, start
        start += 1
        end = start + num_steps
        
    return False, -1
```

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+    finds the last overlapping node and rolls the cycle so it begins with that
+    node. Returns an empty tuple if the cycle does not overlap with the path.
+    '''
+    rolled = ()
```
Maybe `None` is a better default?

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+    path=list(path)
+    for rc in rolled_cycles:
+        path.insert(path.index(rc[0]), rc)
+    path_with_cycles = tuple(path)
+    return path_with_cycles
```
```suggestion
    return  tuple(path)
```

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+def insert_cycles(path, rolled_cycles):
+    '''Inserts already-rolled cycles as a tuple into a path
+    '''
+    path=list(path)
```
Since we pass by reference, does this change `path` in other scopes?  Does it matter?

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+    pathways that could contain 1 or more cycles
+    '''
+    pathways_with_cycles = set()
+    pathways = list(pathways) # turn into list, which supports indexing
```
It doesn't look like you rely on indexing anymore...

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+def find_paths_with_source(pathways, source):
+    '''returns a subset of pathways that contain a given facility as the source
+    '''
+    subset_pathways = set()
+    for path in pathways:
+        if path[0] == source:
+            subset_pathways.add(path)
+
```
Not always sure that this is more clear, but list comprehensions can be useful in cases like this
```suggestion
    subset_pathways = set([ path for path in pathways if path[0] == source])
```

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+def find_paths_with_sink(pathways, sink):
+    '''returns a subset of pathways that contain a given facility as the sink
+    '''
+    subset_pathways = set()
```
ibid

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+    if not facilities:
+        return set()
+
+    p = set()
```
another place for a list comprehension?

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+    if not facilities:
+        return set()
+
+    p = set()
+    for path in pathways:
+        contains = [path for facility in facilities if facility in path]
+        if contains:
+            p.add(contains[0])
```
would set intersection make more sense here?
```suggestion
    facilities = set(facilities)
    p = set( [ path for path in pathways if set(path).intersection(facilities)  ] )
```

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+    '''Finds the pathway with the shortest number of steps from source to
+    target. Returns a tuple with path and length.
+    '''
+    if len(pathways) is not 0:
```
Should you do this check in every method?

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+
+
+def get_shortest_path(pathways):
+    '''Finds the pathway with the shortest number of steps from source to
```
```suggestion
    '''Finds the set of pathways with the shortest number of steps from source to
```

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+        shortest_length = min([len(path) for path in pathways])
+        shortest = set([path for path in pathways if len(path) == shortest_length])
+    else:
+        shortest = "No pathways found"
```
Probably best not to send back some string to be parsed
```suggestion
        shortest = set()
```
Could also be `None`

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+        longest_length = max([len(path) for path in pathways])
+        longest = set([path for path in pathways if len(path) == longest_length])
+    else:
+        longest = "No pathways found"
```
ibid

nuclearkatie replied:
```suggestion
        longest = set()
```

gonuke reviewed (changes requested):
A variety of suggestions and comments in the main code.  I didn't review tests yet, but thought we could get through these changes first

nuclearkatie commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+        infinite_flow = True
+        for edge in edges:
+            if 'capacity' not in H_sg.get_edge_data(*edge):
+                return math.inf
```
I guess networkx already raises a exception if you try to pass it edges with infinite or no capacity, so I probably don't need to introduce my own exceptions right?

nuclearkatie commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+def find_paths_with_sink(pathways, sink):
+    '''returns a subset of pathways that contain a given facility as the sink
+    '''
+    subset_pathways = set()
+    for path in pathways:
+        if path[-1] == sink:
+            subset_pathways.add(path)
+
```
```suggestion
    subset_pathways = set([ path for path in pathways if path[-1] == sink])

```

nuclearkatie commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,319 @@
+    if G.is_multigraph():
+        H = nx.DiGraph(G)
+        if sorted(H.edges()) == sorted(G.edges()):
+            multiedges = 0
```
changed as requested

gonuke reviewed (changes requested):
A couple of streamlining improvements & I'll review tests

nuclearkatie commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,300 @@
+        paths = list(nx.node_disjoint_paths(G, s, t))
+        [ndp.add(tuple(path)) for path in paths]
+    else:
+        print('source and/or target not in graph G!')
```
Changed to raising exception, tests updated accordingly

nuclearkatie commented:
@gonuke I think this PR is waiting on a review on tests and then it will be good to go

gonuke commented on tests/test_pathway_analysis.py:
```diff
@@ -0,0 +1,586 @@
+    assert obs == set()
+
+
+@pytest.fixture
```
Do you use this?

nuclearkatie replied:
```suggestion
```

gonuke commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,295 @@
+
+def get_shortest_path(pathways):
+    '''Finds the set of pathways with the shortest number of steps from source to
+    target. Returns a tuple with path and length.
```
I can't recall if we discussed this already, and whether I am contradicting earlier advice, but I'm not convinced that these should return both a length & the set of shortest paths.  The length can be easily computed from the paths if there are any, and it simplifies the interface.

nuclearkatie replied:
Took out the return of lengths

gonuke reviewed (changes requested):
A few little comments - great testing!

nuclearkatie commented on tests/test_pathway_analysis.py:
```diff
@@ -0,0 +1,586 @@
+
+
+def test_has_multiedges_true():
+    exp = 1
```
```suggestion
    exp = True
```

nuclearkatie commented on tests/test_pathway_analysis.py:
```diff
@@ -0,0 +1,586 @@
+    (obs_H, obs_safe) = pa.transform_to_digraph(G)
+
+    assert obs_safe == exp
+
```
```suggestion


```

nuclearkatie commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,295 @@
+def check_if_sublist(path, list_of_steps):
+    '''Checks to see if a pathway contains each element in steps (list)
+    in order. Returns True/False and the position where the steps begin. If
+    False, returns position -1.
```
took out the True/False and just return pos

nuclearkatie commented on trailmap/pathway_analysis.py:
```diff
@@ -0,0 +1,295 @@
+        return None
+
+
+def transform_to_digraph(G):
```
Added a check for digraph and now returns the original graph G and True if that's the case

nuclearkatie commented:
> A few little comments - great testing!

Thanks @gonuke! I implemented all your requests and addressed your comments about returning extraneous info

gonuke commented:
Thanks for your patience @nuclearkatie
